File: tutorials/mpi/examples/mpi_bug6.f
C******************************************************************************
C FILE: mpi_bug6.f
C DESCRIPTION:
C   This program attempts to compare isend/irecv with send/irecv, but there is
C   a bug. Under AIX it will cause program failure. Intel Linux ignores it. 
C   4 tasks are required.
C SOURCE: Blaise Barney 
C LAST REVISED: 04/02/05
C*****************************************************************************/

      program mpi_bug6
      include 'mpif.h'

      integer REPS, DISP, numtasks, rank, buf, tag, i, tmp,
     +        ierr, dest, src, offset, nreqs
      parameter (DISP=100)
      parameter (REPS=1000)
      integer reqs(REPS*2), stats(MPI_STATUS_SIZE,REPS*2) 
      real*8 T1, T2

      tag = 1

      call MPI_INIT(ierr)
      call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)
      call MPI_COMM_SIZE(MPI_COMM_WORLD, numtasks, ierr)

C     Require 4 tasks
      if (rank .EQ. 0) then
        if (numtasks .NE. 4) then
          print *, 'ERROR: Number of tasks must be 4. Quitting.'
          call MPI_ABORT(MPI_COMM_WORLD, ierr)
        endif
        print *, 'Starting isend/irecv send/irecv test...'
      endif

C     Use barriers for clean output
      call MPI_BARRIER(MPI_COMM_WORLD, ierr)
      print *, 'Task',rank,'starting...'
      call MPI_BARRIER(MPI_COMM_WORLD, ierr)

C     Start the clock
      T1 = MPI_WTIME()
      
C     Tasks 0 and 1 do the isend/irecv test.
C     Determine who to send/receive with. nreqs specifies how many non-blocking
C     operation request handles to capture. offset is where the task should
C     store each request as it is captured in the reqs() array. 
      if (rank .LT. 2) then
        nreqs = REPS*2
        if (rank .EQ. 0) then
          src = 1
          offset = 1
        else
          src = 0
          offset = REPS
        endif
        dest = src

C       Do the non-blocking send and receive operations
        do i=1,REPS 
          call MPI_ISEND(rank, 1, MPI_INTEGER, dest, tag, 
     +                   MPI_COMM_WORLD, reqs(offset), ierr)
          call MPI_IRECV(buf, 1, MPI_INTEGER, src, tag, 
     +                   MPI_COMM_WORLD, reqs(offset+1), ierr)
          offset = offset + 2
          tmp = MOD(i, DISP)
          if (tmp .EQ. 0) then
            print *, 'Task',rank,'has done',i,'isends/irecvs'
          endif
        enddo
      endif

C     Tasks 2 and 3 do the send/irecv test. 
C     Determine who to send/receive with. nreqs specifies how many non-blocking
C     operation request handles to capture. offset is where the task should
C     store each request as it is captured in the reqs() array. 
      if (rank .GT. 1) then
        nreqs = REPS

C       Task 2 does the blocking send operation
        if (rank .EQ. 2) then
          dest = 3
          do i=1,REPS
            call MPI_SEND(rank, 1, MPI_INTEGER, dest, tag,
     +                    MPI_COMM_WORLD, ierr)
            tmp = MOD(i, DISP)
            if (tmp .EQ. 0) then
              print *, 'Task',rank,'has done',i,'sends'
            endif
          enddo
        endif

C       Task 3 does the non-blocking receive operation
        if (rank .EQ. 3) then
          src = 2
          offset = 1
          do i=1,REPS
            call MPI_IRECV(buf, 1, MPI_INTEGER, src, tag,
     +                   MPI_COMM_WORLD, reqs(offset), ierr)
            offset = offset + 1
            tmp = MOD(i, DISP)
            if (tmp .EQ. 0) then
              print *, 'Task',rank,'has done',i,'irecvs'
            endif
          enddo
        endif

      endif

C     Wait for all non-blocking operations to complete and record time
      call MPI_WAITALL(nreqs, reqs, stats, ierr)
      T2 = MPI_WTIME()
      call MPI_BARRIER(MPI_COMM_WORLD, ierr)

      print *, 'Task',rank,'time(wall)=',T2-T1,' sec'

      call MPI_FINALIZE(ierr)

      end

File: mpi/examples/mpi_bug6.c
/******************************************************************************
* FILE: mpi_bug6.c
* DESCRIPTION:  
*   This program attempts to compare isend/irecv with send/irecv, but there is
*   a bug. Under AIX it will cause program failure. Intel Linux ignores it.
*   4 tasks are required.
* AUTHOER: Blaise Barney 
* LAST REVISED: 04/13/05
******************************************************************************/
#include "mpi.h"
#include <stdio.h>
#include <stdlib.h>

#define COMM MPI_COMM_WORLD
#define REPS 1000
#define DISP 100

int main (int argc, char *argv[])
{
int numtasks, rank, buf, tag1=1, i, rc, dest, src, offset, nreqs;
double T1, T2;
MPI_Request reqs[REPS*2];
MPI_Status stats[REPS*2];

MPI_Init(&argc,&argv);
MPI_Comm_size(COMM, &numtasks);
MPI_Comm_rank(COMM, &rank);

/* Require 4 tasks */
if (rank == 0 ) {
  if (numtasks != 4) {
    printf("ERROR: Number of tasks must be 4. Quitting.\n");
    MPI_Abort(COMM, rc);
    }
  printf("Starting isend/irecv send/irecv test...\n");
  }

/* Use barriers for clean output */
MPI_Barrier(COMM);
printf("Task %d starting...\n", rank);
MPI_Barrier(COMM);

T1 = MPI_Wtime();     /* start the clock */

/* Tasks 0 and 1 do the isend/irecv test. 
*  Determine who to send/receive with. nreqs specifies how many non-blocking
*  operation request handles to capture. offset is where the task should
*  store each request as it is captured in the reqs() array.         */
if (rank < 2) {
  nreqs = REPS*2;
  if (rank == 0) {
    src = 1;
    offset = 0;
    }
  if (rank == 1) {
    src = 0;
    offset = REPS;
    }
  dest = src;

/* Do the non-blocking send and receive operations */
  for (i=0; i<REPS; i++) {
    MPI_Isend(&rank, 1, MPI_INT, dest, tag1, COMM, &reqs[offset]);
    MPI_Irecv(&buf, 1, MPI_INT, src, tag1, COMM, &reqs[offset+1]);
    offset += 2;
    if ((i+1)%DISP == 0)
      printf("Task %d has done %d isends/irecvs\n", rank, i+1);
    }
  }

/* Tasks 2 and 3 do the send/irecv test. 
   Determine who to send/receive with. nreqs specifies how many non-blocking
   operation request handles to capture. offset is where the task should
   store each request as it is captured in the reqs() array.  */
if (rank > 1) {
  nreqs = REPS;

/* Task 2 does the blocking send operation */
  if (rank == 2) {
    dest = 3;
    for (i=0; i<REPS; i++) {
      MPI_Send(&rank, 1, MPI_INT, dest, tag1, COMM);
      if ((i+1)%DISP == 0)
        printf("Task %d has done %d sends\n", rank, i+1);
      }
    }

/* Task 3 does the non-blocking receive operation */
  if (rank == 3) {
    src = 2;
    offset = 0;
    for (i=0; i<REPS; i++) {
      MPI_Irecv(&buf, 1, MPI_INT, src, tag1, COMM, &reqs[offset]);
      offset += 1;
      if ((i+1)%DISP == 0)
        printf("Task %d has done %d irecvs\n", rank, i+1);
      }
    }

  }

/* Wait for all non-blocking operations to complete and record time */
MPI_Waitall(nreqs, reqs, stats);
T2 = MPI_Wtime();     /* end time */
MPI_Barrier(COMM);

printf("Task %d time(wall)= %lf sec\n", rank, T2-T1);

MPI_Finalize();
}
